extensions/common/crypto/jwt-verifiable-credentials/src/main/java/org/eclipse/edc/verifiablecredentials/jwt/JwtPresentationVerifier.java

```diff
@@ -43,82 +48,110 @@
      * Verifies the presentation by checking the cryptographic integrity as well as the presence of mandatory claims in the JWT.
      *
      * @param serializedJwt The serialized JWT presentation to be verified.
      * @return A Result object representing the verification result, containing specific error messages in case of failure.
      */
-    public Result<Void> verifyPresentation(String serializedJwt) {
+    @Override
+    public Result<Void> verify(String serializedJwt, VerifierContext context) {
 
         // verify the "outer" JWT, i.e. the VP JWT
-        var vpResult = jwtVerifier.verify(serializedJwt, thisAudience);
-        if (vpResult.failed()) {
-            return vpResult;
+        var audience = context.getAudience();
+        var verificationResult = jwtVerifier.verify(serializedJwt, audience);
+        if (verificationResult.failed()) {
+            return verificationResult;
         }
 
         // verify all "inner" VC JWTs
         try {
-            // obtain the actual VP JSON structure
-            var signedVpJwt = SignedJWT.parse(serializedJwt);
-            var vpClaim = signedVpJwt.getJWTClaimsSet().getClaim(VP_CLAIM);
-            if (vpClaim == null) {
-                return Result.failure("VP-JWT does not contain mandatory claim: " + VP_CLAIM);
+            // obtain the actual JSON structure
+            var signedJwt = SignedJWT.parse(serializedJwt);
+            if (isCredential(signedJwt)) {
+                return verificationResult;
+            }
+
+            if (!isPresentation(signedJwt)) {
+                return Result.failure("Either '%s' or '%s' claim must be present in JWT.".formatted(VP_CLAIM, VC_CLAIM));
             }
+
+            var vpClaim = signedJwt.getJWTClaimsSet().getClaim(VP_CLAIM);
             var vpJson = vpClaim.toString();
 
             // obtain the "verifiableCredentials" object inside
             var map = objectMapper.readValue(vpJson, Map.class);
             if (!map.containsKey(VERIFIABLE_CREDENTIAL_JSON_KEY)) {
                 return Result.failure("Presentation object did not contain mandatory object: " + VERIFIABLE_CREDENTIAL_JSON_KEY);
             }
-            var vcTokens = extractCredentials(map.get(VERIFIABLE_CREDENTIAL_JSON_KEY));
+            var credentialTokens = extractCredentials(map.get(VERIFIABLE_CREDENTIAL_JSON_KEY));
 
-            if (vcTokens.isEmpty()) {
+            if (credentialTokens.isEmpty()) {
                 // todo: this is allowed by the spec, but it is semantic nonsense. Should we return failure or not?
-                return Result.failure("No credential found in presentation.");
+                return Result.success();
             }
 
             // every VC is represented as another JWT, so we verify all of them
-            for (String token : vcTokens) {
-                vpResult = vpResult.merge(jwtVerifier.verify(token, signedVpJwt.getJWTClaimsSet().getIssuer()));
+            for (String token : credentialTokens) {
```

Comment:
I would have call it `rawCredentials` instead of `credentialTokens` to be more explicit but it's just detail :)

Reply:
yup, sounds good also.